torch/csrc/distributed/autograd/context/container.h

```diff
@@ -91,6 +91,18 @@ class TORCH_API DistAutogradContainer {
   static int64_t currentContextId();
 
  private:
+  // Structure holding one shard of the sharded autograd context map.
+  struct ContextsShard {
+    // Map storing autograd contexts for this shard.
+    std::unordered_map<int64_t, ContextPtr> contexts;
+
+    // Lock for this shard.
+    mutable std::mutex lock;
```

Comment:
- consider reordering so that lock is first (acquiring the lock will fetch the beginning part of the unordered_map<>  into cache in exclusive mode automatically)

- consider padding slightly, to avoid interference.
    in fbcode, sizeof() this struct should be 96 bytes. If the hardware cache line is 128, two adjacent entries will contend with each other.
   std::hardware_destructive_interference_size is c++17-only, but could always just add something like the following to the bottom, and call it a day:
       int64_t unusedCachePadding_[8];  // prevent adjacent shards from sharing a cache line

Reply:
oh, so I'm learning that the more modern way is something like

```
 struct alignas(128) ContextsShard {
```